File: android-app/src/main/java/kz/optimabank/optima24/utility/TLSSocketFactory.java
package kz.optimabank.optima24.utility;

import java.io.IOException;
import java.net.InetAddress;
import java.net.Socket;
import java.net.UnknownHostException;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;

import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;

/**
 * @author fkrauthan
 */
public class TLSSocketFactory extends SSLSocketFactory {
    private final SSLSocketFactory internalSSLSocketFactory;

    public TLSSocketFactory(SSLSocketFactory context) throws KeyManagementException, NoSuchAlgorithmException {
        /*SSLContext context = SSLContext.getInstance("TLS");
        context.init(null, null, null);*/
        internalSSLSocketFactory = context;
    }

    @Override
    public String[] getDefaultCipherSuites() {
        return internalSSLSocketFactory.getDefaultCipherSuites();
    }

    @Override
    public String[] getSupportedCipherSuites() {
        return internalSSLSocketFactory.getSupportedCipherSuites();
    }

    @Override
    public Socket createSocket() throws IOException {
        return enableTLSOnSocket(internalSSLSocketFactory.createSocket());
    }

    @Override
    public Socket createSocket(Socket s, String host, int port, boolean autoClose) throws IOException {
        return enableTLSOnSocket(internalSSLSocketFactory.createSocket(s, host, port, autoClose));
    }

    @Override
    public Socket createSocket(String host, int port) throws IOException {
        return enableTLSOnSocket(internalSSLSocketFactory.createSocket(host, port));
    }

    @Override
    public Socket createSocket(String host, int port, InetAddress localHost, int localPort) throws IOException {
        return enableTLSOnSocket(internalSSLSocketFactory.createSocket(host, port, localHost, localPort));
    }

    @Override
    public Socket createSocket(InetAddress host, int port) throws IOException {
        return enableTLSOnSocket(internalSSLSocketFactory.createSocket(host, port));
    }

    @Override
    public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort) throws IOException {
        return enableTLSOnSocket(internalSSLSocketFactory.createSocket(address, port, localAddress, localPort));
    }

    private static Socket enableTLSOnSocket(Socket socket) {
        if (socket != null && (socket instanceof SSLSocket)//&& isTLSServerEnabled((SSLSocket) socket)
                ) { // skip the fix if server doesn't provide there TLS version
            ((SSLSocket) socket).setEnabledProtocols(new String[]{"TLSv1.2"});
        }
        return socket;
    }

    /*private static boolean isTLSServerEnabled(SSLSocket sslSocket) {
        System.out.println("__prova__ :: " + sslSocket.getSupportedProtocols().toString());
        for (String protocol : sslSocket.getSupportedProtocols()) {
            if (protocol.equals("TLS_v1_1")) {
                return true;
            }
        }
        return false;
    }*/

    /*private Socket enableTLSOnSocket(Socket socket) {
        if(socket != null && (socket instanceof SSLSocket)) {
            ((SSLSocket)socket).setEnabledProtocols(new String[] {"TLSv1.1", "TLSv1.2"});
        }
        return socket;
    }*/
}

File: mpp-library/core/data/network/src/androidMain/kotlin/kg/optima/mobile/network/client/TLSSocketFactory.kt
package kg.optima.mobile.network.client

import java.io.IOException
import java.net.InetAddress
import java.net.Socket
import javax.net.ssl.SSLSocket
import javax.net.ssl.SSLSocketFactory


class TLSSocketFactory(private val internalSSLSocketFactory: SSLSocketFactory) : SSLSocketFactory() {
	override fun getDefaultCipherSuites(): Array<String> {
		return internalSSLSocketFactory.defaultCipherSuites
	}

	override fun getSupportedCipherSuites(): Array<String> {
		return internalSSLSocketFactory.supportedCipherSuites
	}

	@Throws(IOException::class)
	override fun createSocket(): Socket {
		return enableTLSOnSocket(internalSSLSocketFactory.createSocket())
	}

	@Throws(IOException::class)
	override fun createSocket(s: Socket, host: String, port: Int, autoClose: Boolean): Socket {
		return enableTLSOnSocket(internalSSLSocketFactory.createSocket(s, host, port, autoClose))
	}

	@Throws(IOException::class)
	override fun createSocket(host: String, port: Int): Socket {
		return enableTLSOnSocket(internalSSLSocketFactory.createSocket(host, port))
	}

	@Throws(IOException::class)
	override fun createSocket(
		host: String,
		port: Int,
		localHost: InetAddress,
		localPort: Int
	): Socket {
		return enableTLSOnSocket(
			internalSSLSocketFactory.createSocket(
				host,
				port,
				localHost,
				localPort
			)
		)
	}

	@Throws(IOException::class)
	override fun createSocket(host: InetAddress, port: Int): Socket {
		return enableTLSOnSocket(internalSSLSocketFactory.createSocket(host, port))
	}

	@Throws(IOException::class)
	override fun createSocket(
		address: InetAddress,
		port: Int,
		localAddress: InetAddress,
		localPort: Int
	): Socket {
		return enableTLSOnSocket(
			internalSSLSocketFactory.createSocket(
				address,
				port,
				localAddress,
				localPort
			)
		)
	}

	companion object {
		private fun enableTLSOnSocket(socket: Socket): Socket {
			if (socket is SSLSocket /*&& isTLSServerEnabled((SSLSocket) socket)*/) { // skip the fix if server doesn't provide there TLS version
				socket.enabledProtocols = arrayOf("TLSv1.2")
			}
			return socket
		}
	}
}